File: algorithms/2_medium/M/00921_minimum_add_to_make_parentheses_valid/o1_space.cpp
class Solution {
public:
    int minAddToMakeValid(string s) {
        int curr_balance{0};
        int min_balance_left{0};
        int min_balance_right{0};

        for (const char& ch : s) {
            if (ch == '(') {
                curr_balance += 1;
            } else {
                curr_balance -= 1;
            }

            min_balance_left = std::min(min_balance_left, curr_balance);
        }

        curr_balance = 0;
        for (auto it{s.rbegin()}; it != s.rend(); ++it) {
            if (*it == ')') {
                curr_balance += 1;
            } else {
                curr_balance -= 1;
            }

            min_balance_right = std::min(min_balance_right, curr_balance);
        }

        return std::abs(min_balance_left) + std::abs(min_balance_right);
    }
};

File: algorithms/2_medium/M/00921_minimum_add_to_make_parentheses_valid/o1_space.rs
impl Solution {
    pub fn min_add_to_make_valid(s: String) -> i32 {
        let mut curr_balance = 0_i32;
        let mut min_balance_left = 0_i32;
        let mut min_balance_right = 0_i32;

        for byte in s.as_bytes().iter() {
            if *byte == b'(' {
                curr_balance += 1;
            } else {
                curr_balance -= 1;
            }

            min_balance_left = min_balance_left.min(curr_balance);
        }

        curr_balance = 0;
        for byte in s.as_bytes().iter().rev() {
            if *byte == b')' {
                curr_balance += 1;
            } else {
                curr_balance -= 1;
            }

            min_balance_right = min_balance_right.min(curr_balance);
        }

        min_balance_left.abs() + min_balance_right.abs()
    }
}
